servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/DefaultHostAndPort.java

```diff
@@ -30,8 +40,71 @@ final class DefaultHostAndPort implements HostAndPort {
      * @param port the port.
      */
     DefaultHostAndPort(String hostName, int port) {
+        this(hostName, port, isIPv6(hostName) ? '[' + hostName + "]:" + port : hostName + ':' + port);
+    }
+
+    private DefaultHostAndPort(String hostName, int port, String toString) {
         this.hostName = requireNonNull(hostName);
         this.port = port;
+        this.toString = requireNonNull(toString);
+    }
+
+    /**
+     * Parse IPv4 {@code xxx.xxx.xxx.xxx:yyyyy} and IPv6 {@code [xxxx:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx]:yyyyy} style
+     * addresses.
+     * @param ipPort An IPv4 {@code xxx.xxx.xxx.xxx:yyyyy} or IPv6
+     * {@code [xxxx:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx]:yyyyy} addresses.
+     * @param startIndex The index at which the address parsing starts.
+     * @return A {@link HostAndPort} where the hostname is the IP address and the port is parsed from the string.
+     */
+    static HostAndPort parseFromIpPort(String ipPort, int startIndex) {
+        String inetAddress;
+        final boolean isv6;
+        int i;
+        if (ipPort.charAt(startIndex) == '[') { // check if ipv6
+            if (ipPort.length() - startIndex > MAX_IPV6_LEN) {
+                throw new IllegalArgumentException("Invalid IPv6 address: " + ipPort);
```

Comment:
Should we use only substring starting from `startIndex` instead of the whole `ipPort` in the exception message?